extras/kms/kms.go

```diff
@@ -0,0 +1,510 @@
+package kms
+
+import (
+	"context"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/hashicorp/go-dbw"
+	wrapping "github.com/hashicorp/go-kms-wrapping/v2"
+	"github.com/hashicorp/go-kms-wrapping/v2/aead"
+	"github.com/hashicorp/go-kms-wrapping/v2/extras/multi"
+	"github.com/hashicorp/go-multierror"
+)
+
+// cachePurpose defines an enum for wrapper cache purposes
+type cachePurpose int
+
+const (
+	// unknownWrapperCache is the default, and indicates that a purpose wasn't
+	// specified
+	unknownWrapperCache cachePurpose = iota
+
+	// externalWrapperCache defines an external wrapper cache
+	externalWrapperCache
+
+	// scopeWrapperCache defines an scope wrapper cache
+	scopeWrapperCache
+)
+
+// Kms is a way to access wrappers for a given scope and purpose. Since keys can
+// never change, only be added or (eventually) removed, it opportunistically
+// caches, going to the database as needed.
+type Kms struct {
```

Comment:
I know it's not our traditional/general pattern, but consider (or feel free not) adding a validity check (possibly in a sync.Once) such that if they instantiate it directly instead of with New any function errors. There aren't nil checks on `k.repo` in the functions so if someone instantiates directly it'll just panic. Alternately you could make the struct private so the only option is via New.

Reply:
There are all sorts of issues with the Kms if they instantiate it directly which is one of the reasons none of its fields are public.  At this point, a Kms has indeterminate behavior if they skip the New factory... so I'm sort of okay with it panic'ing.